modules/hivemq-edge-module-modbus/src/main/java/com/hivemq/edge/adapters/modbus/ModbusProtocolAdapter.java

```diff
@@ -218,21 +205,62 @@ private void calculateDelta(@NotNull final ModBusData modBusData, @NotNull final
         }
     }
 
-    protected @NotNull ModBusData readRegisters(final @NotNull ModbusToMqttMapping modbusToMqttMapping) {
-        try {
-            final AddressRange addressRange = modbusToMqttMapping.getAddressRange();
-            final DataPoint dataPoint = modbusClient.readHoldingRegisters(addressRange.startIdx,
-                    addressRange.nrRegistersToRead, modbusToMqttMapping.getDataType());
-            final ModBusData data = new ModBusData(modbusToMqttMapping);
-            data.addDataPoint(dataPoint);
-            return data;
-        } catch (final Exception e) {
-            throw new RuntimeException(e);
+    protected static @NotNull CompletableFuture<ModBusData> readRegisters(
+            final @NotNull ModbusToMqttMapping modbusToMqttMapping,
+            final @NotNull ModbusClient modbusClient) {
+        final AddressRange addressRange = modbusToMqttMapping.getAddressRange();
+
+        return doRead(addressRange.startIdx, addressRange.nrRegistersToRead, addressRange.unitId, modbusToMqttMapping.getDataType(), addressRange.readType, modbusClient)
+                .thenApply(dataPoint -> {
+                    final ModBusData data = new ModBusData(modbusToMqttMapping);
+                    data.addDataPoint(dataPoint);
+                    return data;
+                });
+    }
+
+    protected static CompletableFuture<DataPoint> doRead(
+            final int startIdx,
+            final int count,
+            final int unitId,
+            final @NotNull ModbusDataType dataType,
+            final @NotNull ModbusAdu readType,
+            final @NotNull ModbusClient modbusClient) {
+        if (HOLDING_REGISTERS.equals(readType)) {
```

Comment:
Wouldnt be a switch statement be a bit more beuatiful?

Reply:
u sure that does not work in java11? I think ot does? 😕

Reply:
it actually does .. not sure what went through my brain when I typed this